consensus/src/aggregation/engine.rs

```diff
@@ -455,30 +466,38 @@ impl<
         };
         let quorum = quorum_from_slice(polynomial);
 
-        // Get the acks
+        // Get the acks and check digest consistency
         let acks_by_epoch = match self.pending.get_mut(&ack.item.index) {
             None => {
                 // If the index is not in the pending pool, it may be confirmed
                 // (i.e. we have a threshold signature for it).
                 return Err(Error::AckIndex(ack.item.index));
             }
             Some(Pending::Unverified(acks)) => acks,
-            Some(Pending::Verified(_, acks)) => acks,
+            Some(Pending::Verified(digest, acks)) => {
+                // If we have a verified digest, ensure the ack matches it
+                if ack.item.digest != *digest {
+                    return Err(Error::AckDigest(ack.item.index));
+                }
+                acks
+            }
         };
 
-        // We already checked that we don't have the ack
-
-        // Add the partial signature
+        // Add the partial signature (if not already present)
         let acks = acks_by_epoch.entry(ack.epoch).or_default();
         if acks.contains_key(&ack.signature.index) {
             return Ok(());
         }
         acks.insert(ack.signature.index, ack.clone());
 
-        // If a new threshold is formed, handle it
-        if acks.len() >= (quorum as usize) {
+        // If there exists a quorum of acks with the same digest, form a threshold signature
+        let filtered_acks = acks
+            .values()
+            .filter(|a| a.item.digest == ack.item.digest) // matches the verified digest (if it exists)
```

Comment:
This was a bug.

Reply:
I think the logic is right but the comment `// matches the verified digest (if it exists)` is wrong. Since it may not be verified. However, the invariant is that there was not previously sufficient `acks` in `partials` to form a threshold signature, there now might be. The only digest for which there may now be enough is for the same digest as the `ack` that just got added. So that's why the logic is correct (though it's not technically verified, it's verified-by-proxy if enough people have signed the ack)

Reply:
I think the wording can be confusing as yes, we are still willing to recover even if there is no verified digest.

Reply:
Updated 👍